Rebase 4890

rainwoodman opened the pull request:
Rebase #5091 to this. Still needs some clean up.

rainwoodman commented:
Note that this version is somehow a lot slower than the one based on scipy/master. (3.25 seconds vs 2.88 seconds)

sturlamolden commented:
Thanks. Let me sleep on this for a couple of hours. 

I see a few things that must be changed, including two bugs. 

The problem with algorithms like kd-tree queries is that small changes can have huge impact on the runtime. But this version should have potential to be faster than SciPy master, so we can probably fix this performance regression.

rainwoodman commented:
1. The slowness was primarily due to the GNU_C optimized distance_p. Swapping between the GNU_C (with goto) and without goto I get 3.29 and 2.96 respectively.

2. The extra p == 2 cases in PR4890 (ckdtreebench) hurts. (2.78). If I swap-in the master version of ckdtree_query.cxx it drops to 2.69, compariable, (may be faster) than scipy/master.

rainwoodman commented:
I am using 

gcc (GCC) 4.9.2 20150212 (Red Hat 4.9.2-6)
Copyright (C) 2014 Free Software Foundation, Inc.

sturlamolden commented:
Ah, then we just take the computed goto version out. It wasn't really a big improvement anyway.

Do you know which p==2. cases it is that hurts? Can you remove them in your branch so I can just merge it in?

Other things:

Can you try to take out the floating point comparisons that checks if a box is unused? Preferably pass NULL instead of a pointer to a valid ckdtrebox. If that is not possible, just use an extra int as a member of ckdtreebox. The latter has the drawback that we must dereference the pointer (as we do now), but we don't do any floating point arithnetics to check if the box is unused. The former is better because it just checks the value of the ckdtreebox pointer itself.

Don't use ```&buf[m] ``` instead of  ```buf + m``` for pointer arithmetics. 

I will fix the ckdtreebox bug that relates to C++ exceptions in Cython. Just leave it as it is. 

The array  ```scratch ``` cannot be automatic. We have to use C++89, not C++14 or C++17, because it has to build on MSVC version 9 (Visual Studio 2003). You can use  ```std::vector ``` instead. This means we have to catch and translate C++ exceptions here as well. You can see how that is done. If you are uncertain just leave it to me and I will fix it.

sturlamolden commented:
Ah, sorry, I didn't see that you removed the floating point comparisons. Looks good now.

rainwoodman commented:
"Do you know which p==2. cases it is that hurts? Can you remove them in your branch so I can just merge it in?" I removed that comment because I immediately realized it wasn't the issue. None of the p==2. case was hurting. The assemblies looked suspecious but after consulting the source code I convinced myself it wasn't an issue.

I thought the compiler always produce the same regard less buf+m or &buf[m], at least in C; but I added that anyways.

You are probably more experienced in exception handling and std::vectors, so I'd prefer to leave those for you to fix.

rainwoodman commented:
For the periodic boundary to make sense, the source data points should be wrapped into a [0, L) box before the tree is built. We shall mention that in the doc string. OK if I add a Notes section to the docstring of query?

sturlamolden commented:
Yes, add all the relevant information to the docstring. And make sure it renders correctly in the SciPy documentation, because it is autogenerated from the docstrings :)

sturlamolden commented:
The optimizer can decide to make &buf[m] become buf+m. But they are actually slightly different. &buf[m] will first dereference the data at buf+m, return this variable, and then we take the address of its return value. I.e. &buf[m] means &(*(buf+m)), not simply buf+m. the difference is that it triggers a read from memory. But the optimizer can decide to skip this read.

rainwoodman commented:
Just thought it makes sense to move boxsize to KDTree.__init__, Then we just add a docstring that when boxsize is not None, the input position and queries are wrapped in a zip([0, 0, ...], boxsize) box.

For other queries that doesn't yet accept a periodic boundary, we will just raise an NotImplemented for now.

rainwoodman commented:
I think that also paves the way to eliminate allocate/free in ckdtreebox. I'll push a commit.

rainwoodman commented:
Seems like the other queries just need to update the distance trackers with min/max awareness. 

How frozen are the PointRectTracker and RectRectTracker code?

sturlamolden commented:
Nothing is frozen until it is merged for SciPy 0.17, which might be as late as December :)

sturlamolden commented:
Yes, it makes sence to havet it already in ```__init__```.

sturlamolden commented:
We don't have to request wrapping in an [0, L) box if we do this already in init. We can just copy the data and do the wrapping if it is outside the box.

sturlamolden commented:
The headerfile that defines the struct ckdtree to C++ must also be updated when you add a new member to cKDTree in Cython.

sturlamolden commented:
You must also update the pickling code to make sure the box is included in the pickle.

sturlamolden commented:
We also need more test cases:
- We must make sure cKDTree works correctly when initialized with a box. 
- We must verify that the pickling code works with and without a box (Python 3 only).

rainwoodman commented:
Copying. What about the cases where the memory is tight? Then we don't want to copy, even if boxsize is set.

Now we have a combinational mess between boxsize and copy_data:

- When copy_data is False, and boxsize is not None:
  raise exception is points are outside of the box
- When copy_data is True, and boxsize is not None:
  wrap points into the box in the copy
- When copy_data is False, and boxsize is None:
  do not copy the data
- When copy_data is True, and boxsize is None:
  copy the data.

By delegating the wrapping to the caller, we don't have do worry about this complex logic. 
ascontinousarray will create a copy when the source array is of wrong type or strides; so the next line would create an extra copy when copy_data is True. This can probably be simplified. (not sure how yet)

- What would be a test case to test cKDTree works `correctly` when initialized with a box? There is already an overall periodic query test to ensure the periodic tree gives the correct K-NN query results. Thus the tree can't be very wrong.

I will push a fix to the pickling code.

rainwoodman commented:
BTW. I found a fix to the Python 2 pickling as well. 

I would like to be sure the tracker's are 'frozen' because to implement periodic boundary on other queries I need to modify them.

sturlamolden commented:
Having a copy_data argument will be confusing. Data are sometimes copied, out of necessity. Setting copy_data=False might give the false impression that this is avoided. I think it is better that we just copy the data. We should perhaps consider to issue a warning when data are copied by cKDTree instead.

sturlamolden commented:
The trackers will not be changed before you are done with them, if that is what you mean. They do what they should now, so I am not touching them.

sturlamolden commented:
The eay to test if cKDTree works correctly with a box is to create a small random data set (with a fixed seed) and compare the query results with brute force search. make sure a part of the data is outside the box. Then do this for all combinations of build flags. (You only need to do the brute force once.)

sturlamolden commented:
Thanks for fixing pickling on Python 2.

rainwoodman commented:
Yes. I am in for removing the copy_data argument; issuing a warning, maybe, maybe not. 

What's the larger landscape in scipy? Do people like the idea of issuing warnings for unnecessary data copies just for type/alignment/stride conversion?

Anyhow, if copy is avoided, then modifying the original data point becomes unfeasible. 
We can do wrapping every time a position is accessed for the periodic case(slow!) 
I still feel it is simpler and faster just raise if the points are not properly wrapped. (a O[N] at the beginning)
Especially, wrapping is just one line: pos = pos % boxsize.

That test is already implemented as test_ckdtree_box; though I based the ground truth on a non-periodic kdtree. Still prefer a brute force comparison?

sturlamolden commented on scipy/spatial/ckdtree/src/query.cxx:
```diff
@@ -432,7 +492,9 @@ query_knn(const ckdtree      *self,
-    npy_intp i;
+    npy_intp i, j;
+    
+    npy_float64 scratch[m];
```
This is a GNU extension. Replace with
```
npy_float64 *scratch = PyMem_Malloc(m * sizeof(npy_float64));
if (NPY_UNLIKELY(scratch == NULL)) return PyErr_NoMemory();
```

Then call 
```
PyMem_Free(scratch);
```
right below the line that says ```NPY_END_ALLOW_THREADS```.

sturlamolden commented:
I figured it is easier to avoid the GNU extension with a malloc than use std::vector which would add another layer of C++ exception handlers. Everything above and below the GIL release is effectively C, so a malloc will be safe.

rainwoodman commented:
I am thinking just test if the query position is in the box and raise an exception if not. Then this wrapping code can go away. (and we will have a consistent behavior requiring all positions to be in the box)

sturlamolden commented:
No, leave it there. Just get rid of it for the non-periodic case. We should not require any data to be inside the box.

sturlamolden commented:
Here is what you should do:

```C++
/* Query n points for their k nearest neighbors */

extern "C" PyObject*
query_knn(const ckdtree      *self, 
          npy_float64        *dd, 
          npy_intp           *ii, 
          const npy_float64  *xx,
          const npy_intp     n, 
          const npy_intp     k, 
          const npy_float64  eps, 
          const npy_float64  p, 
          const npy_float64  distance_upper_bound,
          const ckdtreebox * box)
{
    /* release the GIL */
    NPY_BEGIN_ALLOW_THREADS
    {
        try {
            const npy_intp m = self->m;        
            if (NPY_LIKELY(box == NULL) {
                /* non-periodic space */
                for (npy_intp i=0; i<n; ++i) {
                    npy_float64 *dd_row = dd + (i*k);
                    npy_intp *ii_row = ii + (i*k);
                    const npy_float64 *xx_row = xx + (i*m);                
                    query_single_point(self, dd_row, ii_row, xx_row, k, eps, p, 
                        distance_upper_bound, ::infinity, box);
                 }
            }
            else {
                /* periodic space */
                std::vector<npy_float64> scratch(m); 
                for (npy_intp i=0; i<n; ++i) {
                    npy_float64 *dd_row = dd + (i*k);
                    npy_intp *ii_row = ii + (i*k);
                    const npy_float64 *xx_row = xx + (i*m);                
                    for (npy_intp j=0; j<m; ++j) {
                        scratch[j] = xx_row[j];
                        while(scratch[j] < 0) 
                            scratch[j] += box->fbox[j];
                        while(scratch[j] >= box->fbox[j]) 
                            scratch[j] -= box->fbox[j];
                    }
                    query_single_point(self, dd_row, ii_row, &scratch[0], k, 
                        eps, p, distance_upper_bound, ::infinity, box);
                }
            } 
        } 
        catch(...) {
            translate_cpp_exception_with_gil();
        }
    }
    /* reacquire the GIL */
    NPY_END_ALLOW_THREADS

    if (PyErr_Occurred()) 
        /* true if a C++ exception was translated */
        return NULL;
    else {
        /* return None if there were no errors */
        Py_RETURN_NONE;
    }
}
```

sturlamolden commented:
@rgommers Just pinging in case you want to follow.

sturlamolden commented:
And instead of
```C++
for (j=0; j<m; ++j) {
    scratch[j] = xx_row[j];
    while(scratch[j] < 0) 
        scratch[j] += box->fbox[j];
    while(scratch[j] >= box->fbox[j]) 
        scratch[j] -= box->fbox[j];
}
```
You can do
```C++
for (j=0; j<m; ++j)
    scratch[j] = std::fmod(xx_row[j], box->fbox[j]);
```

sturlamolden commented:
(If fmod rounds similar to Python -- it might not, I have to check)

sturlamolden commented:
```
#include <math.h>
#include <stdio.h>
int main(int argc, char *argv[])
{
    printf("%f\n", fmod(4.5,2));
    printf("%f\n", fmod(-4.5,2));
    return 0;
}
```

Gives
```
$ ./test
0.500000
-0.500000
```

Bugger, it rounds towards zero. Ok, then we do this:

```C++
for (j=0; j<m; ++j) {
    const npy_float64 box->fbox[j];
    scratch[j] = std::fmod(xx_row[j], tmp);
    if (scratch[j] < 0) scratch[j] += tmp;
}
```

rainwoodman commented:
Yes. It makes sense to do wrapping in the query function. I'll put it back in.

Someone did some benchmark over floor, fmod and a while-loop.

http://stackoverflow.com/a/11980755

Based on that result, I propose we use this: (some benchmarks?) 

int i = 0
for(i=0; x >= boxsize; i ++) {
if(i== 3) x = x - floor(x / boxsize) * boxsize;
x-= boxsize;
}
for(i=0; x <0; i ++) {
if(i== 3) x = x - floor(x / boxsize) * boxsize;
x+= boxsize;
}

Anyways, this is an implementation detail and we can settle on it later.

sturlamolden commented:
Those benchmarks are actually ok. You can see what happens if the distance is large. Also, it is not just a matter of performance, but also accuracy. We want to avoid accumulating rounding error. I don't think the difference between using ```std::floor``` or ```std::fmod``` is very important. The main thing is to avoid excessive additions or subtractions.

sturlamolden commented:
Yes, put the wrapping back in and set it up approximately or exactly as I did. You can also see where to put the ```scratch``` array to make sure C++ exceptions are caught. Don't use VLA because it is a GNU extension (will fail on MSVC).

sturlamolden commented:
How is porting of the other methods and the trackers going? I was going to write a ```query_radius```  method, but I will hold that back until you are finished. I am also planning on writing a ball-tree for ```scipy.spatial``` (faster to search than kd-tree), but I will hold that back as well. 

Another optimization which I have in my head is to template the ```_distance_p``` function. It is not that important for id-trees because we are limited to Minkowski distances, but ball-trees allows arbitrary distance functions. For kd-trees it would bloat the machine code by a factor of four, but we would get rid of the checks on the value of ```p``.

rainwoodman commented:
I did lift the comparison outside of the loop. I suspect most compiler probably will be able to lift that expression outside of the loop -- at least I know craycc does this and sometimes do it wrongly :) 

While coding this up, I noticed you used i*m for the offsets. Is there a reason not using ii_row += m? (pulling out from my memory + is quit a lot faster than *) Avoiding data dependency? Since this loop is driving a much slower kd look up, it probably doesn't matter much. 

The min_distance_p functions. I feel quite unease about their current forms. There are too many of them!
I will introduce a periodic version for each, for now, just to get it working. We can clean it up later.

sturlamolden commented:
Stylistically (cf. NumPy coding standards) it should be like this

```C++
inline const npy_float64 
_wrap(const npy_float64 x, const npy_float64 box) 
{
    const npy_float64 r = std::floor(x / box);
    return x - r * box;
}
```

sturlamolden commented:
As for the pointer artihmetics, it is integer math. Speedwise there is nothing to gain form using ```ii += k``` instead of ```ii_row = ii + (i*k)``` because integer math is so fast. Because this computations happens so rarely, it is not obvious that the former will be faster.

sturlamolden commented:
As for the ```min_distance_p``` functions. Those are just copied from the Cython code that Pat Varilly wrote. I haven't given them much thought. I just ported them directly from Cython to C++. Yes there are many of them and probably plenty of DRY violations.

sturlamolden commented:
Sorry for nagging about NumPy coding standards. But we have to follow as much as we can. The only exception is when differences between C and C++ make it impossible.

rainwoodman commented:
I am thinking this may be the place to make use of a template that passes in a set of min_max functions -- one for periodic, one for non-periodic. 

The ckdtree object needs to be coupled into this (for boxsize elements). Hopefully this new interface can be useful for a ball tree as well.

rainwoodman commented:
So I played around; does this look like valid C++? (it compiled with g++)

```
struct A {
    static inline void f() {

    }
};
struct B {
    static inline void f() {
    }
};

template<typename C> void func() {
    C::f();
}

int main() {

    func<A>();
    func<B>();
}
```

sturlamolden commented:
Yes, it does.

rgommers commented:
> @rgommers Just pinging in case you want to follow.

Thanks Sturla. I'm following from a distance. And quite liking the collaboration dynamics here!

sturlamolden commented on scipy/spatial/ckdtree/src/query.cxx:
```diff
@@ -329,79 +355,113 @@ query_single_point(const ckdtree *self,
+            // set up children for searching
+            // inf2 will be pushed to the queue
+            if (NPY_LIKELY(self->raw_boxsize_data == NULL)) {
+                memcpy(inf2->side_distances(), inf->side_distances(), sizeof(npy_float64) * ( m)); 
```
std::memcpy

add std:: to all other libc calls as well

sturlamolden commented on scipy/spatial/ckdtree/src/rectangle.h:
```diff
@@ -61,158 +61,191 @@ struct Rectangle {
+            } else {
+                /* min below, max above */
+                *realmax = half;
+                *realmin = fmin(min, full - max);
```
dmin

rainwoodman commented:
I think this is mostly done, functionality-wise. 

There must be issues with C++ and the compiler but I am not an expert with and I would say better if @sturlamolden takes over from this point.

I still would like to look into these
- see if there is a way to clean up and improve test_kdtree.py; this can be a separate PR.
- rebase K-NN query to fully make use of MinMaxDist; that will get rid of a few more branchings but I worry about a performance regression. This one needs more thoughts, discussion and experiments.

sturlamolden commented:
Ok, just tell me when you want me to pull this in. It might be that it needs a rebase after I have recovered the lost commits after the rebase mishap.

rainwoodman commented:
Just realized query_ball_point was missing the wrapping part. 

What is the rebase about? Looks like this is a fast forward from the cpp_kdtree branch.

sturlamolden commented:
I was an idiot and lost some commits when rebasing. So when I recover them, this will lag behind some commits. But it should be easy to fix.

sturlamolden commented:
Just take the time you need. We still have about three months before 0.17 is branched.

rainwoodman commented:
Sure. I just don't want to be blocking you from improving other things. What about you pull this in, rebase, and I file another PR for porting the query_knn to MinMaxDist based on the new head?

rainwoodman commented:
1. Overall the template based code is faster than 882416076 by 10% to 20%.

2. I do get some errors in Running spatial.Neighbors.time_sparse_distance_matrix in 882416076 and here.

sturlamolden commented:
The errors in ```sparse_distance_matrix``` might be my fault. The rebase error happened after I fixed a bug which was carried over from the old Cython code.

sturlamolden commented:
What are those errors?

rainwoodman commented:
I remember they are in the distance calculation. Related to sparse matrices. It also fails a test case there , probably related.

rainwoodman commented:
As an excuse to further procrastinate my real work, I will refactor MinMaxDist to give each p-norm distance a template. That will get rid of more branching in the main loop.

nodeinfo_pool looks similar to the stack in Tracker. Can we somehow merge those two? then query kNN will look exactly the same as other algorithms.

sturlamolden commented:
If you do, the code to keep is ```nodeinfo_pool```. It is optimized for fast allocation, by preallocating pages. It ensures that ```nodeinfo``` structs are aligned for fast access. It uses RAII to ensure cleanup of ```nodeinfo``` structs, and it makes sure that no deallocation happens in the middle of the query. It is perhaps the most important hand-tuned optimization in ```cKDTree```, so be careful not to replace it with something less efficient.

sturlamolden commented:
Also, the stack in the tracker serves the same purpose as the priority queue in query, whereas ```nodeinfo_pool``` is just the allocator.

rainwoodman commented:
My first attempt failed. 

Here is what I tried:

- MinMaxDist<T>, MinMaxDistInf<T>  
- T can be MinMaxDist1D and MinMaxDistBox1D for non-periodic and periodic 1D distance between point-point and rect-rect. 

The logic works to some extent, but somehow the periodic case all failed. Maybe it shall be done more carefully. 

I will take another stab. Trying to merge p=inf and p=2 with function override (special type for square and inf).

rainwoodman commented:
Why was a normal distribution used in the test cases?

sturlamolden commented:
p=inf is not a compile-time constant and cannot be special cased.

rainwoodman commented:
You are right.

In the end I lifted the comparisons up,  into the query_xxx functions. This avoids most of the expensive branchings checking p==1,2,inf during distance tracking. Thus, the code runs slightly faster. It was a fun experiment but I am OK to revert to the previous version, especially if there are concerns about readability of the new code. I liked the idea of the core algorithm (push) not branch on p, but I dislike templates.

- distance.h and distance_box.h and rectangle.h defines a bunch of interfaces for Minkowski distances, which are (mostly, but Pinf) simple sums of Dist1D. 

- I am hoping templates do not make the code harder to decipher, do they?

- Also notice how the special optimized code for p==2 non-periodic is placed as an override near the end of distance.h.

- The 'HANDLE' macro simplifies the branching code in all of the cxx drivers, making the lines
  look like a dispatch table. If macros are to be avoided, they can be expanded, but notice 
  we are looking at 8 different cases (4 x 2), so it can get quite verbose.

-  The current code passes all tests except the one about recarray data type, though I am still wary about the power of these test cases.

sturlamolden commented:
I haven't looked at the current code, but you have my blessing to complicate the code as much as we need. 

The recarry issue is probably just a missing ```aligned=True``` in the call to ```np.dtype``` (cf. the ```coo_entries``` class in ```ckdtree.pyx```), which I think was one of the commits I accidentally lost. Just add that and see what happens.

rainwoodman commented:
Nope. What helped was adding a line in 1143, handling the 'recarray' special case.

rainwoodman commented:
Any ideas how to merge Rectangle with nodeinfo?

sturlamolden commented:
Conceptually they are related, but I am not sure it's worth it.

rainwoodman commented:
I will leave them as is then.

sturlamolden commented:
Should I pull this as is now? I have been a bit busy the last weeks.

rainwoodman commented:
Sure. Let's do that and regroup our efforts later, when we both have some time. 

Is there a bullet lists of what to be done before merging back to scipy?

sturlamolden commented:
Hi
I have been a bit busy, and will be this week as well. But it's not forgotten :)

sturlamolden commented:
Ok, I am pulling this now since it could be done automatically. But I will not have a chance to work on #4890 before next week (I think).

sturlamolden commented:
Thanks Yu Feng @rainwoodman :)

rainwoodman commented:
Thanks for merging. I will make myself available to help out.

sturlamolden commented:
It did build fine, but we got a minor test failure on ```sparse_distance_matrix```. Not to worry, it should be easy to fix :)

sturlamolden commented:
Yes, in the test suite the ouput type was spelled ```recarray``` instead of ```ndarray```. That is probably my own mistake :)

sturlamolden commented:
Also, there is a segfault on Python 3.4 and 3.3, and a failure on building the documentation. Probably just minor things.

rainwoodman commented:
Did you take look at the segfaults? any suggestions on reproducing them?

sturlamolden commented:
I am just terribly busy atm. Looks like it's going to last until the first week of November. I will have to prioritise but this is on my TODO list.
